pkg/server/singleprocess/service_pipeline.go

```diff
@@ -0,0 +1,234 @@
+func (s *Service) GetPipeline(
+	ctx context.Context,
+	req *pb.GetPipelineRequest,
+) (*pb.GetPipelineResponse, error) {
+	if err := serverptypes.ValidateGetPipelineRequest(req); err != nil {
+		return nil, err
+	}
+
+	p, err := s.state(ctx).PipelineGet(req.Pipeline)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the graph for the steps so we can get the root. We enforce a
+	// single root so the root is always the first step.
+	stepGraph, err := serverptypes.PipelineGraph(p)
+	if err != nil {
+		return nil, err
+	}
+
+	orderedStep := stepGraph.KahnSort()
+	rootStepName := orderedStep[0].(string)
+
+	return &pb.GetPipelineResponse{
+		Pipeline: p,
+		RootStep: rootStepName,
+		// TODO: Leaving this out intentionally for now, need to convert stepGraph into mermaid
+		//Graph:    graph,
+	}, nil
+}
+
+func (s *Service) ListPipelines(
+	ctx context.Context,
+	req *pb.ListPipelinesRequest,
+) (*pb.ListPipelinesResponse, error) {
+	if err := serverptypes.ValidateListPipelinesRequest(req); err != nil {
+		return nil, err
+	}
+
+	result, err := s.state(ctx).PipelineList(req.Project)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.ListPipelinesResponse{
+		Pipelines: result,
+	}, nil
+}
+
+func (s *Service) RunPipeline(
+	ctx context.Context,
+	req *pb.RunPipelineRequest,
+) (*pb.RunPipelineResponse, error) {
+	log := hclog.FromContext(ctx)
+	if err := serverptypes.ValidateRunPipelineRequest(req); err != nil {
+		return nil, err
+	}
+
+	// Get the pipeline we should execute
+	pipeline, err := s.state(ctx).PipelineGet(req.Pipeline)
+	if err != nil {
+		return nil, err
+	}
+
+	// Generate job IDs for each of the steps. We need to know the IDs in
+	// advance to setup the dependency chain.
+	stepIds := map[string]string{}
+	for name := range pipeline.Steps {
+		stepIds[name], err = server.Id()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Generate the jobs for each of the steps
+	var stepJobs []*pb.QueueJobRequest
+	for name, step := range pipeline.Steps {
+		var dependsOn []string
+		for _, dep := range step.DependsOn {
+			dependsOn = append(dependsOn, stepIds[dep])
+		}
+
+		job := proto.Clone(req.JobTemplate).(*pb.Job)
+		job.Id = stepIds[name]
+		job.DependsOn = append(job.DependsOn, dependsOn...)
+
+		// Queue the right job depending on the Step type. We will queue a Waypoint
+		// operation if the type is a reserved built in step.
+		switch o := step.Kind.(type) {
+		case *pb.Pipeline_Step_Build_:
+			job.Operation = &pb.Job_Build{
+				Build: &pb.Job_BuildOp{
+					DisablePush: o.Build.DisablePush,
+				},
+			}
+		case *pb.Pipeline_Step_Deploy_:
+			job.Operation = &pb.Job_Deploy{
+				Deploy: &pb.Job_DeployOp{},
+			}
+
+			if o.Deploy.Release {
+				// TODO(briancain): do it
+				// copy `job` and update Operation to be release I think. then append
+				// job to stepJobs
+
+				// Queue a release job too
+				log.Warn("Currently not queueing a release job yet....sry!!!")
```

Comment:
I left this unimplemented because I'm not sure if it makes sense in the pipelines view. If someone wants to do a release, they'd just add a Release step....